### ISSUE TITLE
Strip out standard AutoHotkey version information

### ISSUE THREAD

**Issue by Lexikos** [AutoHotkey/Ahk2Exe#90]
> I assume Ahk2Exe only sets the version information if specified by directives. I think this should be changed for files based on an exe, to strip the standard AutoHotkey version information. Normally the bin files omit this information.
> [(I posted in May 2021)](https://www.autohotkey.com/boards/viewtopic.php?p=399694#p399694)

Specifically, I think the file description (e.g. "AutoHotkey 32-bit") and product name ("AutoHotkey") should be stripped out of compiled scripts. Version number is fine; the bin files include it.

There might be a use case for customizing a base file and then using it for different scripts, in which case the user might want these fields left as is. I suppose that could be supported by checking if the value matches known standard patterns before stripping it out. I think that would be `^AutoHotkey ((ANSI|Unicode) )?(32|64)-bit$`  at the moment, but it doesn't need to be that precise. (AutoHotkey_H adds "_H" to the file description and product name, but I'm not sure whether the standard compiler even works for it.)

**Comment by TAC109** [comment]
Currently an .exe base file must have a file description containing the word  `AutoHotkey` (as well as an appropriate file version number) in order to be recognised as a [valid base file](https://www.autohotkey.com/docs/Scripts.htm#information). Would it be sufficient to just always blank the file description and product name prior to processing any relevant compiler directives?

(There has been no attempt to make the standard compiler compatible with AutoHotkey_H.)

**Comment by joedf** [comment]
Perhaps @HotKeyIt could shed some light on that note?

**Comment by TAC109** [comment]
I’ll be away from home for a week, and will resume this after I get back.

**Comment by Lexikos** [comment]
@TAC109 
I think what you're saying is that because a file with some other description would not have been recognized as a valid base file in the first place, it's already known that the description is "AutoHotkey" something, and there's no need to take into account the "customizing" scenario I described.

But I'm guessing that what you say only applies to discovery of exe files in the installation directory. A custom base file would almost certainly be somewhere else, and be specified explicitly via the GUI or command line, or a directive.

I did some testing for a custom scenario and learned the following:
  - Any base file selected by the user via the GUI (and maybe command line?) must have a filename beginning with "AutoHotkey". The restriction is neither necessary nor helpful when the user is explicitly selecting a specific file as base. It is probably a side-effect of code sharing and not specifically intended to apply to this situation. It is harmful in the case where the "base file invalid" message is missed, which is likely given that it only appears in the status bar.
  - The file description is **not** required to start with "AutoHotkey". In my test, it was "1".

Specifically, the scenario I tested was for embedding **multiple scripts** into one exe, where one of the scripts sets the product name/description. I used two methods, and both worked:
  1. Compile with a standard base file with resource `#2`, then rename the result and use it as the base file to compile `#1` with the product name/description directives.
  2. As above but in reverse order: `#1` with directives, then `#2`.

I think the file description and product name can be left as they are if the resource ID is not the default one. If there is no `#1` resource, the file is essentially still AutoHotkey.

> Would it be sufficient to just always blank the file description and product name prior to processing any relevant compiler directives?

Is that *always* blank the file description and product name prior to *the stage at which* compiler directives *would be processed* if they are present? Or is it blank the file description and product name only if relevant compiler directives are found, and doing it prior to processing them? If there are no compiler directives, prior to processing relevant directives could mean not doing it at all, depending on your meaning.

<hr/>

**In summary**, I think it would be best to set a default blank product name and description if and only if **both** of the following are true:
  - The resource ID is `#1` (whether it was specified or left as default).
  - The base file's current product name is "AutoHotkey".

I assumed before that the values set by the directives are simply recorded and only acted upon at a later stage, so simply setting default values (under those conditions) would be sufficient to make the change happen. Having a brief look at the code seems to have affirmed my assumption - it looks like just a case of setting `state.verInfo.ProductName` and `state.verInfo.FileDescription` by default. Any directives in the script would implicitly override the default values.

**Comment by TAC109** [comment]
@Lexikos Thanks for your detailed testing. 

While the method you used for compiling multiple embedded scripts can be used for ad hoc compiling, I had envisioned that this type of compile would most likely be accomplished by using the AddResource compiler directive to specify additional scripts to Ahk2Exe. The .exe can then be created simply with just one pass of the compiler, and any possibly error-prone fiddling with intermediate base files is avoided. (Script files that are processed by the AddResource directive are not restricted in any way, and can make use of the full facilities of the compiler such as #Include processing, etc.)

Regarding point 2 of your summary; I have coded the "AutoHotkey" test against the File Description field. Do you wish this changed to the Product Name field as stated? (I couldn’t see any discussion in the main body of your post regarding this.)

As v1 AutoHotkey version info includes the additional fields of `InternalName` `LegalCopyright` and `OriginalFilename` I am removing these as well. (Incidentally I noticed that the LegalCopyright field has a terminating year of 2013.)

I should have these changes pushed to GitHub in the next few days.

**Comment by TAC109** [comment]
I've pushed my Ahk2Exe amendments for this to GitHub. Attached is the compiled version.

[Ahk2Exe.zip](https://github.com/AutoHotkey/Ahk2Exe/files/8934714/Ahk2Exe.zip)

**Comment by Lexikos** [comment]
> ... using the AddResource compiler directive to specify additional scripts to Ahk2Exe. 

 >Script files that are processed by the AddResource directive are not restricted in any way, and can make use of the full facilities of the compiler such as #Include processing, etc.

Are scripts added that way preprocessed in the same way as the main script, including all `@Ahk2Exe` directives? That is useful, but also surprising and potentially unwanted. Ideally one can opt in/out via the directive. At the very least, it should be documented that `.ahk` resources are given special treatment.

> I have coded the "AutoHotkey" test against the File Description field. 

That's fine.

> (I couldn’t see any discussion in the main body of your post regarding this.)

I edited it too many times. Considering a file that does something related to AutoHotkey but isn't actually part of AutoHotkey, I presumed it would be more likely to contain "AutoHotkey" in the file description (e.g. "SciTE4AutoHotkey" or "AutoHotkey-based widget") than for the product name to be exactly "AutoHotkey". Along those lines, comparing the product name directly could be both simpler and more accurate than a partial match of the description. On the other hand, AutoHotkey itself might hypothetically install a tool that is part of the "AutoHotkey" product but isn't an interpreter.

Given that this is only to determine whether to remove the version information, `BinType.Description ~= "^AutoHotkey"` is fine.

**Comment by TAC109** [comment]
I have adjusted the AddResource documentation to describe how added scripts are processed.
